gnovm/pkg/test/test.go

```diff
@@ -0,0 +1,478 @@
+func (tp *testParams) SetBytes(key string, val []byte)  { /* noop */ }
+func (tp *testParams) SetInt64(key string, val int64)   { /* noop */ }
+func (tp *testParams) SetUint64(key string, val uint64) { /* noop */ }
+func (tp *testParams) SetString(key string, val string) { /* noop */ }
+
+// ----------------------------------------
+// main test function
+
+// TestOptions is a list of options that must be passed to [Test].
+type TestOptions struct {
+	// BaseStore / TestStore to use for the tests.
+	BaseStore storetypes.CommitStore
+	TestStore gno.Store
+	// Gno root dir.
+	RootDir string
+	// Used for printing program output, during verbose logging.
+	Output io.Writer
+	// Used for os.Stderr, and for printing errors.
+	Error io.Writer
+
+	// Not set by NewTestOptions:
+
+	// Flag to filter tests to run.
+	RunFlag string
+	// Whether to update filetest directives.
+	Sync bool
+	// Uses Error to print when starting a test, and prints test output directly,
+	// unbuffered.
+	Verbose bool
+	// Uses Error to print runtime metrics for tests.
+	Metrics bool
+	// Uses Error to print the events emitted.
+	Events bool
+
+	filetestBuffer bytes.Buffer
+	outWriter      proxyWriter
+}
+
+// WriterForStore is the writer that should be passed to [Store], so that
+// [Test] is then able to swap it when needed.
+func (opts *TestOptions) WriterForStore() io.Writer {
+	return &opts.outWriter
+}
+
+// NewTestOptions sets up TestOptions, filling out all "required" parameters.
+func NewTestOptions(rootDir string, stdin io.Reader, stdout, stderr io.Writer) *TestOptions {
+	opts := &TestOptions{
+		RootDir: rootDir,
+		Output:  stdout,
+		Error:   stderr,
+	}
+	opts.BaseStore, opts.TestStore = Store(
+		rootDir, false,
+		stdin, opts.WriterForStore(), stderr,
+	)
+	return opts
+}
+
+// proxyWriter is a simple wrapper around a io.Writer, it exists so that the
+// underlying writer can be swapped with another when necessary.
+type proxyWriter struct {
+	w io.Writer
+}
+
+func (p *proxyWriter) Write(b []byte) (int, error) {
+	return p.w.Write(b)
+}
+
+// tee temporarily appends the writer w to an underlying MultiWriter, which
+// should then be reverted using revert().
+func (p *proxyWriter) tee(w io.Writer) (revert func()) {
+	save := p.w
+	if save == io.Discard {
+		p.w = w
+	} else {
+		p.w = io.MultiWriter(save, w)
+	}
+	return func() {
+		p.w = save
+	}
+}
+
+func Test(
+	memPkg *gnovm.MemPackage,
+	fsDir string, // directory on filesystem of package; used for Sync
+	opts *TestOptions,
+) error {
+	opts.outWriter.w = opts.Output
+
+	var errs error
+
+	// Stands for "test", "integration test", and "filetest".
+	tset, itset, itfiles, ftfiles := parseMemPackageTests(opts.TestStore, memPkg)
+
+	// testing with *_test.gno
+	if len(tset.Files)+len(itset.Files) > 0 {
+		// create a common cw/gs for both the `pkg` tests as well as the `pkg_test`
+		// tests. this allows us to "export" symbols from the pkg tests and
+		// import them from the `pkg_test` tests.
```

Comment:
```suggestion
		// Import them from the `pkg_test` tests.
```

Reply:
did not apply this one, as it's a continuation of the above sentence